src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -3191,6 +3209,44 @@ String MergeTreeData::getPartitionIDFromQuery(const ASTPtr & ast, ContextPtr loc
     return partition_id;
 }
 
+DataPartsVector MergeTreeData::getDataPartsVector(ContextPtr local_context) const
+{
+    return getVisibleDataPartsVector(local_context->getCurrentTransaction());
+}
+
+MergeTreeData::DataPartsVector MergeTreeData::getVisibleDataPartsVector(const MergeTreeTransactionPtr & txn) const
```

Comment:
const references to shared pointers are kinda weird...

On a more serious note, this should probably take a 'Snapshot' and not a 'Transaction'. 'Transaction' is for writing, and 'Snapshot' is for reading, and you might have one w/o another. Not sure how important this is at this experimental stage...

Reply:
Maybe yes, but then we should add current transaction id into 'Snapshot' to allow transaction see its own changes (or we can pass 'Snapshot' and optional 'TransactionID').

P.S. PR isn't ready for review yet, I'm just trying to run some tests in CI. Code is ugly and I'm going to refactor it after making tests pass.

Reply:
OK, I'll stop for now, ping me if you need more comments.